mmdet/models/detectors/dab_detr.py

```diff
@@ -0,0 +1,652 @@
+        query_embed = query_embed.unsqueeze(1).repeat(
+            1, bs, 1)  # [num_query, 4] -> [num_query, bs, 4]
+        masks = masks.view(bs, -1)  # [bs, h, w] -> [bs, h*w]
+        memory = self.encoder(
+            query=feat, query_pos=pos_embed, query_key_padding_mask=masks)
+        if self.num_patterns == 0:
+            target = torch.zeros(
+                self.num_query, bs, self.embed_dims, device=query_embed.device)
+        else:
+            target = self.patterns.weight[:, None, None, :]\
+                .repeat(1, self.num_query, bs, 1)\
+                .view(-1, bs, self.embed_dims)
+            query_embed = query_embed.repeat(self.num_patterns, 1, 1)
+
+        out_dec, reference = self.decoder(
+            query=target,
+            key=memory,
+            key_pos=pos_embed,
+            query_pos=query_embed,
+            key_padding_mask=masks,
+            reg_branches=reg_branches)
+        if return_memory:
+            memory = memory.permute(1, 2, 0).reshape(bs, c, h, w)
+            return out_dec, reference, memory
+        return out_dec, reference, None
+
+    def _forward(self,
+                 batch_inputs: Tensor,
+                 batch_data_samples: OptSampleList = None) -> Tuple[Tensor]:
+        """Network forward process.
+
+        Includes backbone, neck and head forward without post-processing.
+        """
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec, reference, _ = self.forward_transformer(
+            **transformer_inputs_dict)
+        results = self.bbox_head.forward(outs_dec, reference)
+        return results
+
+    def loss(self, batch_inputs: Tensor,
+             batch_data_samples: SampleList) -> Union[dict, list]:
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec, reference, _ = self.forward_transformer(
+            **transformer_inputs_dict)
+        losses = self.bbox_head.loss(outs_dec, reference, batch_data_samples)
+        return losses
+
+    def predict(self,
+                batch_inputs: Tensor,
+                batch_data_samples: SampleList,
+                rescale: bool = True) -> SampleList:
+        img_feats = self.extract_feat(batch_inputs)
+        transformer_inputs_dict = self.forward_pretransformer(
+            img_feats, batch_data_samples)
+        outs_dec, reference, _ = self.forward_transformer(
+            **transformer_inputs_dict)
+        results_list = self.bbox_head.predict(
+            outs_dec, reference, batch_data_samples, rescale=rescale)
+        batch_data_samples = self.add_pred_to_datasample(
+            batch_data_samples, results_list)
+        return batch_data_samples
+
+
+class ConditionalAttention(BaseModule):
```

Comment:
We can discuss the order of each module in the file. Personally, I think it is top-down, so it should be decoder first, then Layer, and finally attention

Reply:
we should talk about ConditionalAttention refactor in next meeting

Reply:
> we should talk about ConditionalAttention refactor in next meeting

No problem my homies.

Reply:
> We can discuss the order of each module in the file. Personally, I think it is top-down, so it should be decoder first, then Layer, and finally attention

No problem my homies.

Reply:
I prefer to move ConditionalAttention to layers/transfomer.py as one of the  base module ,we can talk about it.

Reply:
> I prefer to move ConditionalAttention to layers/transfomer.py as one of the base module ,we can talk about it.

facts.